### PR TITLE
Adding a new pair style in SPIN package 

### PR DESCRIPTION
**Pull request by julient31**
**Summary**

- Adding a new pair interaction in the SPIN package (exchange/biquadratic).
- Adding an option to offset the energy and mechanical force in the exchange and exchange/biquadratic pair styles.
- Correcting error in force calculation in the pair/spin neel interaction.
- Reformatting the if conditions in the pair/spin interactions.
- Small correction of the constants in the dipole/spin pair interactions.
- Improving the test problems for the exchange interactions.

**Related Issue(s)**

None

**Author(s)**

Julien Tranchida, Sandia Labs
jtranch@sandia.gov

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

- the new option for the pair_style spin/exchange cannot be called in former versions.
- for runs containing _pair spin/exchange_ interactions, older restart files will no longer be compatible. 

**Implementation Notes**

All validation tests (src/SPIN/test_problems) were succesfully rerun.
Changes are limited to the SPIN package.

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Comment by julient31** [comment]
Hello Axel @akohlmey and Richard @rbberger !
Hope you guys are doing well. 

I was reviewing the modified files of that PR, and realized that I did not remove some changes I made in atom_mask.h.

Maybe a quick explanation about that: 
Stan and I are working on creating a spin/kk style to make the spin calculations kokkos compatible (for now no speed-up of the magnetic interactions, but speed-up of the associated mechanical interactions, for example Snap). I started to work on this in this branch, but moved those modifications to another one (spin-kokkos). 
I realized I forgot to remove the mask I had to create in atom_mask.h for those changes. They are useless for the changes in that PR, but should be useful soon for the PR of the spin-kokkos branch, when it is ready. 

Is it fine if I leave those changes of atom_mask.h in this PR? Or would you rather like me to remove them?

**Comment by akohlmey** [comment]
@julient31 no worries. I noticed the change and then had to look up what it was used for and realized that it is a KOKKOS thing. So it is down to @stanmoore1 to decide.

**Comment by julient31** [comment]
Hello everyone,
Is there anything I can do to help with this PR?
Best,
Julien.

**Comment by akohlmey** [comment]
> Hello everyone,
> Is there anything I can do to help with this PR?

This was added too late to be included in the upcoming stable release, so it has been on hold since.

File: src/SPIN/pair_spin_dmi.cpp [hunk 1]
```diff
@@ -244,31 +244,36 @@ void PairSpinDmi::compute(int eflag, int vflag)
...
+        if (eflag) {
+          evdwl -= (spi[0]*fmi[0] + spi[1]*fmi[1] + spi[2]*fmi[2]);
+          evdwl *= 0.5*hbar;
+          emag[i] += evdwl;
+        } else evdwl = 0.0;
+        
+        f[i][0] += fi[0];
+        f[i][1] += fi[1];
+        f[i][2] += fi[2];
+          if (newton_pair || j < nlocal) {
```
Comment (rbberger) [inline src/SPIN/pair_spin_dmi.cpp:260]: @julient31 there's some weird indentation here

Comment (julient31) [reply src/SPIN/pair_spin_dmi.cpp]: Indeed, good catch. Thanks Richard.

**Comment by stanmoore1** [comment]
I'm fine with leaving in the `atom_masks.h` change, but I don't think it will work with Kokkos as is because the value for some masks is the same as others. I.e. `SP_MASK` is the same value as `X_MASK`.

**Comment by julient31** [comment]
> I'm fine with leaving in the `atom_masks.h` change, but I don't think it will work with Kokkos as is because the value for some masks is the same as others. I.e. `SP_MASK` is the same value as `X_MASK`.

Hey Stan! Thanks for checking. I left the same values as in the spin-kokkos branch. I guess we need to correct them over there. 
Could it explain the temperature difference (as the spin temperature is computed with the values of fm, and I just check that the fm values are different by a factor 2 between serial and kokkos_mpi_only)?

**Comment by stanmoore1** [comment]
The Kokkos atom_masks are only used on the GPU, so it shouldn't affect `kokkos_mpi_only`.

**Comment by rbberger** [comment]
@sjplimp Please have a look at this one.

**Review by sjplimp — APPROVED** [review]

File: src/SPIN/pair_spin_exchange.cpp [hunk 2]
```diff
@@ -507,10 +560,12 @@ void PairSpinExchange::read_restart_settings(FILE *fp)
 {
   if (comm->me == 0) {
     utils::sfread(FLERR,&cut_spin_exchange_global,sizeof(double),1,fp,nullptr,error);
+    utils::sfread(FLERR,&e_offset,sizeof(int),1,fp,nullptr,error);
```
Comment (rbberger) [inline src/SPIN/pair_spin_exchange.cpp:563]: This has backward compatibility implications. Older restarts won't be read correctly.

Comment (julient31) [reply src/SPIN/pair_spin_exchange.cpp:563]: Yes, this is correct Richard. I added it in the backward compatibility section of the PR notice.

File: src/SPIN/compute_spin.cpp [hunk 3]
```diff
@@ -148,10 +148,12 @@ void ComputeSpin::compute_vector()
   int i;
   int countsp, countsptot;
   double mag[4], magtot[4];
+  double m2, m2tot;
+  double m4, m4tot;
   double magenergy, magenergytot;
   double tempnum, tempnumtot;
   double tempdenom, tempdenomtot;
-  double spintemperature;
+  double spintemperature,binder;
```
Comment (rbberger) [inline src/SPIN/compute_spin.cpp:156]: unused variables

**Review by rbberger — COMMENTED** [review]
@julient31 I've merged the current master into this branch and did some minor corrections. While reviewing the code I noticed that there are several changes to files outside of exchange, exchange/biquadratic that aren't mentioned in the PR description. While those seem to minor corrections, you should mention why those were necessary.

File: src/SPIN/compute_spin.cpp [hunk 4]
```diff
@@ -148,10 +148,12 @@ void ComputeSpin::compute_vector()
   int i;
   int countsp, countsptot;
   double mag[4], magtot[4];
+  double m2, m2tot;
+  double m4, m4tot;
```
Comment (rbberger) [inline src/SPIN/compute_spin.cpp:152]: unused variables

Comment (julient31) [reply src/SPIN/compute_spin.cpp]: Hey Richard,
I removed the declaration of those quantities (binder,m2,m2tot,m4,m4tot). I do not use them in compute/spin anymore.

**Review by rbberger — APPROVED** [review]

**Pull request merged**